mm2src/coins/lp_coins.rs

```diff
@@ -1487,6 +1506,29 @@ pub trait CoinAssocTypes {
     fn parse_signature(&self, sig: &[u8]) -> Result<Self::Sig, Self::SigParseError>;
 }
 
+/// Defines associated types specific to each coin (Pubkey, Address, etc.)
+pub trait NftAssocTypes {
+    type TokenContractAddr: Send + Sync + fmt::Display;
+    type TokenContractAddrParseError: fmt::Debug + Send + fmt::Display;
+    type TokenId: ToBytes + Send + Sync;
+    type TokenIdParseError: fmt::Debug + Send + fmt::Display;
+    type Chain: ToBytes + Send + Sync;
+    type ChainParseError: fmt::Debug + Send + fmt::Display;
+    type ContractType: ToBytes + Send + Sync;
+    type ContractTypeParseError: fmt::Debug + Send + fmt::Display;
+
+    fn parse_token_contract_address(
+        &self,
+        token_contract_addr: &[u8],
+    ) -> Result<Self::TokenContractAddr, Self::TokenContractAddrParseError>;
+
+    fn parse_token_id(&self, token_id: &[u8]) -> Result<Self::TokenId, Self::TokenIdParseError>;
+
+    fn parse_chain(&self, chain: &[u8]) -> Result<Self::Chain, Self::ChainParseError>;
+
+    fn parse_contract_type(&self, contract_type: &[u8]) -> Result<Self::ContractType, Self::ContractTypeParseError>;
+}
```

Comment:
Will we actually need those in swaps code? I am not sure yet, parsing data is used for p2p data and from storage data so maybe we will need for the from storage data but for p2p data I think the TokenContractAddr/TokenId/Chain/ContractType will be known for both sides before starting the swap. We can leave them for now but we have to be careful with adding coin specific logic inside swap state machine logic which we did in legacy swap and want to avoid in v2 code to make it cleaner.

Reply:
I see two options of nft swap implementation:
1. Have optional nft fields in current args structures, for example in `SendMakerPaymentArgs`. So in maker state changes we will need to check do we have Some in nft related fields or None. So in each maker state we will have to handle two inner implementations: for coin and for NFT. We may have the same problem as we had during trying to integrate nft activation into existing coin related code, which will lead to decision of logic separation.

2. Create separate NFT related structures for args (eg `SendNftMakerPaymentArgs`) and NFT Maker states for `MakerSwapStateMachine`. This way we wont need to check optional nft fields in args as all of them will be required. Also we wont need to handle the coin and nft behaviors in current `on_changed` functions for coin maker states, as we will need to implement `NFT related states`. I believe we still have same `struct MakerSwapStateMachine`.

> I think the TokenContractAddr/TokenId/Chain/ContractType will be known for both sides before starting the swap

I wasnt sure about it so just decided to prepare the functionality. As we already use `parse_...`  functions in `maker_swap_v2.rs` and `taker_swap_v2.rs` I decided that we may have nft related fields in `args` structures or `negotiation` data, so we need this.


The another big approach of nft implementation is providing in `args structure`(doesnt matter for now in NFT args or in optional nft field in current `ags` structures, its about having or not having `NftAssocTypes`) the string of concatenated "token_address,token_id". 
This way Maker coin should be NFT global "coin" from the beginning I think, from its `"nfts_infos"` field we can get `NftInfo` value using "token_address,token_id" key. You know, this approach sounds not bad. This way we wont need `NftAssocTypes`. Need to refactor NFT activation and enable global NFT using nft json from coin config file.

Reply:
> Create separate NFT related structures for args (eg SendNftMakerPaymentArgs) and NFT Maker states for MakerSwapStateMachine

NFT states for swap state machines is a nice idea if it can be done cleanly :)

Reply:
> I wasn't sure about it so just decided to prepare the functionality.

We can leave it in this PR and re-assess when implementing NFT functionalities in the swaps state machine.